apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,216 @@
+          .slice(0, 19),
+        endDate: adjustedDate(accessRule.assessment_access_rule.end_date)
+          .toISOString()
+          .slice(0, 19),
+        active: accessRule.assessment_access_rule.active,
+        credit: parseInt(accessRule.assessment_access_rule.credit),
+        timeLimitMin: parseInt(accessRule.assessment_access_rule.time_limit_mins),
+        password: accessRule.assessment_access_rule.password,
+        examUuid: accessRule.assessment_access_rule.exam_uuid,
+      };
+      const filteredRules = Object.fromEntries(Object.entries(rule).filter(([_, value]) => value));
+      return filteredRules;
+    });
+    assessmentAccessRulesInput.value = JSON.stringify(accessRulesMap);
+    form.submit();
+  });
+
+  function adjustedDate(dateString: string | Date) {
+    const date = new Date(dateString);
+    const timezoneOffset = date.getTimezoneOffset() * 60000;
+    return new Date(date.getTime() - timezoneOffset);
+  }
+
+  function handleUpdateAccessRule({ form }: { form: HTMLFormElement }) {
+    const formData = new FormData(form);
+    const updatedAccessRules: Record<string, any> = Object.fromEntries(formData);
+    const row = parseInt(updatedAccessRules.row.toString());
+    updatedAccessRules.uids = updatedAccessRules.uids
+      .toString()
+      .split(',')
+      .map((uid: string) => uid.trim());
+    updatedAccessRules.active = updatedAccessRules.active === 'true';
+
+    updatedAccessRules.start_date = new Date(
+      Temporal.PlainDateTime.from(updatedAccessRules.start_date)
+        .toZonedDateTime(timezone)
+        .toInstant().epochMilliseconds,
+    ).toISOString();
+
+    updatedAccessRules.end_date = new Date(
+      Temporal.PlainDateTime.from(updatedAccessRules.end_date)
+        .toZonedDateTime(timezone)
+        .toInstant().epochMilliseconds,
+    ).toISOString();
+
+    if (accessRulesData[row]) {
+      accessRulesData[row].assessment_access_rule = updatedAccessRules;
+    } else {
+      accessRulesData[row] = { assessment_access_rule: updatedAccessRules };
+    }
+    refreshTable();
+  }
+
+  on('click', '.editButton', (e) => {
+    const editButton = (e.target as HTMLElement).closest('button');
+    if (!editButton) return;
+
+    templateFromAttributes(editButton, editAccessRuleModal as HTMLElement, {
+      'data-row': '.access-rule-row',
+      'data-title-text': '.modal-title',
+      'data-submit-text': '.updateAccessRuleButton',
+      'data-access-rule-mode': '.access-rule-mode',
+      'data-access-rule-uids': '.access-rule-uids',
+      'data-access-rule-start-date': '.access-rule-start-date',
+      'data-access-rule-end-date': '.access-rule-end-date',
+      'data-access-rule-active': '.access-rule-active',
+      'data-access-rule-credit': '.access-rule-credit',
+      'data-access-rule-time-limit': '.access-rule-time-limit',
+      'data-access-rule-password': '.access-rule-password',
+      'data-access-rule-exam-uuid': '.access-rule-exam-uuid',
+    });
```

Comment:
It seems like we're maintaining some state in the DOM and other state in memory. I'm going to guess that you dump everything into data attributes so that you can use `templateFromAttributes`, is that right? What if we instead provided an API like the following:

```js
templateFromData(editButton, {
  '.access-rule-row': accessRule.index,
  '.access-rule-uids': accessRule.uids
  // ...
})
```

That is, an API that allows you to template from an object, not just a DOM element.

That said, I also wonder why we don't just construct the modal DOM on the fly by calling `EditAccessRuleModal`?

Reply:
I have updated to use the last recommendation here. As of [660f81b](https://github.com/PrairieLearn/PrairieLearn/pull/9934/commits/660f81bebe37d51fcef94b1535961f4abfa02d88), we are now calling `EditAccessRuleModal` on the fly and updating the DOM.